File: bindings/swift/Tests/ZtokTests/RwkvTests.swift
// RwkvTests — RWKV "World" tokenizer coverage. Mirrors the Python
// binding's tests/test_rwkv.py: loads the real rwkv_vocab_v20230424.txt
// fixture (skipped when absent) and checks ztok reproduces the canonical
// reference encodings, matching the in-tree gate in src/rwkv_world.zig.
//
// All cases skip via `try XCTSkipUnless(libztokAvailable)` when the
// shared library can't be loaded, and via XCTSkip when the vocab fixture
// is not present.

import XCTest

@testable import Ztok

final class RwkvTests: XCTestCase {

    // Golden id sequences captured from BlinkDL's canonical reference
    // tokenizer (see bench/rwkv_parity.py / src/rwkv_world.zig).
    private static let golden: [(String, [UInt32])] = [
        ("Hello, world!", [33155, 45, 40213, 34]),
        ("emoji 😀🚀✨ test",
         [34295, 33, 3319, 153, 129, 3319, 155, 129, 10059, 32223]),
        ("0 1 2 10 99 100", [49, 284, 285, 3483, 3572, 3483, 49]),
    ]

    override func setUpWithError() throws {
        try XCTSkipUnless(TestSupport.libztokAvailable,
                          "libztok not loadable (set LD_LIBRARY_PATH or build with `zig build`)")
    }

    /// Locate `bench/vocabs/rwkv_vocab_v20230424.txt` relative to this
    /// source file: Tests/ZtokTests/RwkvTests.swift -> ... -> repo root.
    private static func vocabPath() -> String {
        // #filePath: <repo>/bindings/swift/Tests/ZtokTests/RwkvTests.swift
        // Drop: RwkvTests.swift, ZtokTests, Tests, swift, bindings -> repo.
        var url = URL(fileURLWithPath: #filePath)
        for _ in 0..<5 {
            url.deleteLastPathComponent()
        }
        return url
            .appendingPathComponent("bench")
            .appendingPathComponent("vocabs")
            .appendingPathComponent("rwkv_vocab_v20230424.txt")
            .path
    }

    private func rwkvPipeline() throws -> Pipeline {
        let path = RwkvTests.vocabPath()
        guard FileManager.default.fileExists(atPath: path) else {
            throw XCTSkip("RWKV vocab fixture not present at \(path)")
        }
        return try Pipeline.fromRwkv(path: path)
    }

    func testMatchesReference() throws {
        let pipe = try rwkvPipeline()
        defer { pipe.close() }
        for (text, want) in RwkvTests.golden {
            XCTAssertEqual(try pipe.encode(text), want, "encode mismatch for \(text)")
        }
    }

    func testRoundTrips() throws {
        let pipe = try rwkvPipeline()
        defer { pipe.close() }
        for (text, _) in RwkvTests.golden {
            let ids = try pipe.encode(text)
            XCTAssertEqual(try pipe.decode(ids), text)
        }
    }
}

File: bindings/dotnet/Ztok.Tests/RwkvTests.cs
// RwkvTests.cs — RWKV "World" tokenizer coverage for the .NET binding.
//
// Loads the real rwkv_vocab_v20230424.txt fixture (skipped when absent)
// and checks ztok reproduces the canonical reference encodings, matching
// the in-tree gate in src/rwkv_world.zig and the Go/Python suites.
//
// Skips cleanly (returns early) when libztok cannot be loaded, matching
// the skip pattern in SmokeTests.

using System;
using System.IO;
using Xunit;
using Ztok;

namespace Ztok.Tests;

public class RwkvTests
{
    private static bool LibAvailable() => SmokeTests.LibAvailable();

    // Golden id sequences captured from BlinkDL's canonical reference
    // tokenizer (see bench/rwkv_parity.py / src/rwkv_world.zig).
    public static readonly (string Text, uint[] Want)[] Golden =
    {
        ("Hello, world!", new uint[] { 33155, 45, 40213, 34 }),
        ("emoji \U0001F600\U0001F680✨ test",
            new uint[] { 34295, 33, 3319, 153, 129, 3319, 155, 129, 10059, 32223 }),
        ("0 1 2 10 99 100", new uint[] { 49, 284, 285, 3483, 3572, 3483, 49 }),
    };

    // Walk up from the test assembly dir to the repo root and locate the
    // RWKV vocab fixture. Returns null if it can't be found.
    private static string? FindVocab()
    {
        var dir = new DirectoryInfo(AppContext.BaseDirectory);
        for (int i = 0; i < 10 && dir is not null; i++)
        {
            var probe = Path.Combine(dir.FullName, "bench", "vocabs", "rwkv_vocab_v20230424.txt");
            if (File.Exists(probe)) return probe;
            dir = dir.Parent;
        }
        return null;
    }

    private static Pipeline? OpenRwkv()
    {
        var vocab = FindVocab();
        if (vocab is null)
        {
            Console.Error.WriteLine("RWKV vocab fixture not present; skipping.");
            return null;
        }
        return Pipeline.FromRwkv(vocab);
    }

    [Fact]
    public void MatchesReference()
    {
        if (!LibAvailable()) return;
        using var pipe = OpenRwkv();
        if (pipe is null) return;
        foreach (var (text, want) in Golden)
        {
            Assert.Equal(want, pipe.Encode(text));
        }
    }

    [Fact]
    public void RoundTrips()
    {
        if (!LibAvailable()) return;
        using var pipe = OpenRwkv();
        if (pipe is null) return;
        foreach (var (text, _) in Golden)
        {
            var ids = pipe.Encode(text);
            Assert.Equal(text, pipe.Decode(ids));
        }
    }

    [Fact]
    public void AutoDetectIsRwkv()
    {
        if (!LibAvailable()) return;
        var vocab = FindVocab();
        if (vocab is null)
        {
            Console.Error.WriteLine("RWKV vocab fixture not present; skipping.");
            return;
        }
        Assert.Equal(Format.Rwkv, ZtokLibrary.DetectFormat(vocab));
        // Open should dispatch to FromRwkv via auto-detect.
        using var pipe = Pipeline.Open(vocab);
        Assert.Equal(new uint[] { 33155, 45, 40213, 34 }, pipe.Encode("Hello, world!"));
    }
}
